Can't compile crate that depends on a crate that depends on a proc_macro crate

Windows 7 x64, with rust msvc compiled at 0172e463b0f0c720069d96602209f77c50380066 as well as rustc 1.15.0-nightly (ba872f270 2016-11-17)

It seems a bin/lib crate that depends on a lib crate that in turn depends on a proc macro crate doesn't compile because rustc complains that the proc macro crate was compiled as a dylib.

I have a proc macro crate `foo`, a lib crate `bar` that depends on `foo`, and a bin crate `baz` that depends on `bar`. `bar` compiles fine and generates `libbar.rlib`, but `baz` fails to compile with:

```
error[E0462]: found staticlib `foo` instead of rlib or dylib which `bar` depends on
  --> src\main.rs:1:1
  |
1 | extern crate bar;
  | ^^^^^^^^^^^^^^^^^
  |
  = help: please recompile that crate using --crate-type lib
  = note: crate `foo` path #1: C:\Stuff\Sources\baz\target\debug\deps\foo-7f298898a71a7d96.dll.lib
```

Repro: https://github.com/Arnavion/rust-37839 Run `cargo build` under `baz/`

---

On Linux, since only a .so is generated (no .lib equivalent), the error is different but the cause is presumably the same.

```
error[E0463]: can't find crate for `foo` which `bar` depends on
 --> src/main.rs:1:1
  |
1 | extern crate bar;
  | ^^^^^^^^^^^^^^^^^ can't find crate
```

Note again that this error is when compiling `baz`. There is no problem when compiling `bar`.

--- a/src/librustc_metadata/creader.rs
+++ b/src/librustc_metadata/creader.rs
@@ -318,11 +318,11 @@ impl<'a> CrateLoader<'a> {
                      name: &str,
                      hash: Option<&Svh>,
                      span: Span,
-                     kind: PathKind,
+                     path_kind: PathKind,
                      mut dep_kind: DepKind)
                      -> (CrateNum, Rc<cstore::CrateMetadata>) {
         info!("resolving crate `extern crate {} as {}`", name, ident);
-        let result = if let Some(cnum) = self.existing_match(name, hash, kind) {
+        let result = if let Some(cnum) = self.existing_match(name, hash, path_kind) {
             LoadResult::Previous(cnum)
         } else {
             info!("falling back to a load");
@@ -332,7 +332,7 @@ impl<'a> CrateLoader<'a> {
                 ident: ident,
                 crate_name: name,
                 hash: hash.map(|a| &*a),
-                filesearch: self.sess.target_filesearch(kind),
+                filesearch: self.sess.target_filesearch(path_kind),
                 target: &self.sess.target.target,
                 triple: &self.sess.opts.target_triple,
                 root: root,
@@ -350,7 +350,7 @@ impl<'a> CrateLoader<'a> {
                 let mut proc_macro_locator = locator::Context {
                     target: &self.sess.host,
                     triple: config::host_triple(),
-                    filesearch: self.sess.host_filesearch(PathKind::Crate),
+                    filesearch: self.sess.host_filesearch(path_kind),
                     rejected_via_hash: vec![],
                     rejected_via_triple: vec![],
                     rejected_via_kind: vec![],

--- a/src/test/run-make/issue-37839/Makefile
+++ b/src/test/run-make/issue-37839/Makefile
@@ -1,0 +1,6 @@
+-include ../tools.mk
+
+all:
+	$(RUSTC) a.rs && $(RUSTC) b.rs
+	$(BARE_RUSTC) c.rs -L dependency=$(TMPDIR) --extern b=$(TMPDIR)/libb.rlib \
+		--out-dir=$(TMPDIR)

--- a/src/test/run-make/issue-37839/a.rs
+++ b/src/test/run-make/issue-37839/a.rs
@@ -1,0 +1,13 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(proc_macro)]
+#![allow(unused)]
+#![crate_type = "proc-macro"]

--- a/src/test/run-make/issue-37839/b.rs
+++ b/src/test/run-make/issue-37839/b.rs
@@ -1,0 +1,12 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type = "lib"]
+#[macro_use] extern crate a;

--- a/src/test/run-make/issue-37839/c.rs
+++ b/src/test/run-make/issue-37839/c.rs
@@ -1,0 +1,12 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type = "lib"]
+extern crate b;

